4.0/en/0x21-V13-API.md

```diff
@@ -56,6 +56,18 @@ Note: Due to issues with XXE attacks against DTDs, DTD validation should not be
 | **13.4.1** | Verify that a query allow list or a combination of depth limiting and amount limiting is used to prevent GraphQL or data layer expression Denial of Service (DoS) as a result of expensive, nested queries. For more advanced scenarios, query cost analysis should be used. |  | ✓ | ✓ | 770 |
 | **13.4.2** | Verify that GraphQL or other data layer authorization logic should be implemented at the business logic layer instead of the GraphQL layer. |  | ✓ | ✓ | 285 |
 
+## V13.5 WebSocket Security Requirements
+
+| # | Description | L1 | L2 | L3 | CWE |
+| :---: | :--- | :---: | :---:| :---: | :---: |
+| **13.5.1** | Verify that only authenticated users can open WebSocket connections. | ✓ | ✓ | ✓ | 306 |
+| **13.5.2** | Verify that WebSocket Secure (wss) is used for all WebSocket connections. | ✓ | ✓ | ✓ | 319 |
+| **13.5.3** | Verify that the Origin header is checked in the initial HTTP WebSocket handshake. | ✓ | ✓ | ✓ | 346 |
+| **13.5.4** | Verify that WebSocket messages are properly sanitized and encoded server-side. | ✓ | ✓ | ✓ | 116 |
```

Comment:
Sanitization and/or encoding should be performed before input is used in a different context, e.g. incorporated into a file path, used in a SQL query, written back to a web page's HTML, etc. It is not clear what the purpose is here.

Do we mean input validation to ensure that messages received conform to expected formats?

Reply:
You are right. What do you think of such a formulation:

> Verify that websocket input from the client is sanitized server-side before being processed

This item is in part a repetition of section 5 of ASVS. The aim is to make sure that Websocket inputs are taken into account, as they are not mentioned in section 5. Maybe it's better to modify section 5 though?

Reply:
I removed the rule altogether, to avoid duplication.